packages/zoe/src/contractFacet/contractFacet.js

```
@@ -0,0 +1,348 @@
+          // unburned assets.
+          zcfSeatAdmin.commit(seatStaging);
+          E(zoeMintP).withdrawAndBurn(totalToBurn);
+        },
+      });
+      return zcfMint;
+    };
+
+    /** @type ContractFacet */
+    const zcf = {
+      reallocate: (/** @type SeatStaging[] */ ...seatStagings) => {
+        // We may want to handle this with static checking instead.
+        // Discussion at: https://github.com/Agoric/agoric-sdk/issues/1017
+        assert(
+          seatStagings.length >= 2,
+          details`reallocating must be done over two or more seats`,
+        );
+
+        seatStagings.forEach(seatStaging =>
+          assert(
+            allSeatStagings.has(seatStaging),
+            details`The seatStaging ${seatStaging} was not recognized`,
+          ),
+        );
+
+        // Ensure that rights are conserved overall. Offer safety was
+        // already checked when an allocation was staged for an individual seat.
+        const flattened = arr => [].concat(...arr);
+        const flattenAllocations = allocations =>
+          flattened(allocations.map(allocation => Object.values(allocation)));
+
+        const previousAllocations = seatStagings.map(seatStaging =>
+          seatStaging.getSeat().getCurrentAllocation(),
+        );
+        const previousAmounts = flattenAllocations(previousAllocations);
+
+        const newAllocations = seatStagings.map(seatStaging =>
+          seatStaging.getStagedAllocation(),
+        );
+        const newAmounts = flattenAllocations(newAllocations);
+
+        assert(
+          areRightsConserved(getAmountMath, previousAmounts, newAmounts),
+          details`Rights are not conserved in the proposed reallocation`,
+        );
+
+        // Commit the staged allocations and inform Zoe of the
+        // newAllocation.
+        seatStagings.forEach(seatStaging =>
+          seatToZCFSeatAdmin.get(seatStaging.getSeat()).commit(seatStaging),
+        );
```

Until #1391 is fixed, please put a comment here linking to that issue.

Immediately above this block, good to note the "commit point" --- the point where all the prior code of the method caused no effects and checked all the conditions which could cause this to fail. After the commit point here (as is typical) all effects must happen atomically.

---

This raises an issue we're not setup to handle well: We should somehow indicate when we have an atomic block of multiple effects, where once we're past the commit point, all effects must happen successfully. If an operation within such a commit block throws an error, that should probably be fatal for the vat incarnation --- potentially causing a rollback to prior consistent state. At a tc39 meeting, @waldemarhorwat suggested something like that (the fatal error, not the rollback) in response to our "OOM Must Fail Fast" proposal. We need something like that. This applies to the ERTP implementation as well.

---

https://github.com/Agoric/agoric-sdk/issues/1501